libs/pilight/protocols/433.92/tfa303221.c

```diff
@@ -0,0 +1,268 @@
+ * 501 235 502 472 256 476 262 226 501 478 260 227 500 473 259 233 502 472 264 227 502 231 498 479
+ * 263 225 499 233 500 478 258 229 2056 19207
+ */
+static int findPayloadStartIndex(void) {
+	if (tfa303221->rawlen >= MIN_RAW_LENGTH && tfa303221->rawlen <= MAX_RAW_LENGTH) {
+		int headerFound = 0;
+		for (int i = 0; i < tfa303221->rawlen; i++) {
+			// both pulse & gap need to be > HEADER_PULSE_GAP_MIN_LENGTH
+			if (tfa303221->raw[i] >= HEADER_PULSE_GAP_MIN_LENGTH) {
+				headerFound++;
+			} else if (headerFound > 0) {
+				if (headerFound == HEADER_PULSE_COUNT) {
+					return i;  // this is the first pulse/gap pair that didn't look like a header
+				} else {
+					return -headerFound;  // header not yet complete - broken transmission
+				}
+			}
+		}
+	}
+	return -999;
+}
+
+static void parseCode(void) {
+	int payloadStartIndex = findPayloadStartIndex();
+	if (payloadStartIndex > 0) {
+		// convert pulse-train to array of 0/1 int values
+		unsigned int buffer[40];
+		for (int i = 0; i < 40; i++) {	// read 40 value-pairs from raw data
+			int v0 = tfa303221->raw[payloadStartIndex + (i * 2)];
+			int v1 = tfa303221->raw[payloadStartIndex + (i * 2) + 1];
+			if (v0 < 357 && v1 > 357) {
+				buffer[i] = 0;
+			} else if (v0 > 357 && v1 < 357) {
+				buffer[i] = 1;
+			} else {
+				return;	 // found invalid v0 / v1 pair, abort parsing
+			}
+		}
+
+		// compare message checksum with computed checksum...
+		int checksum = binToDecRev(buffer, 32, 39);
+
+		uint8_t raw[4];
+		raw[0] = binToDecRev(buffer, 0, 7);	   // 4 bytes checksum
+		raw[1] = binToDecRev(buffer, 8, 15);   // 4 bytes checksum
+		raw[2] = binToDecRev(buffer, 16, 23);  // 4 bytes checksum
+		raw[3] = binToDecRev(buffer, 24, 31);  // 4 bytes checksum
+
+		int crc = lfsr_digest8_reflect(raw, 4, 0x31, 0xf4);	 // 4 bytes checksum
+
+		if (crc != checksum) {
+			return;
+		}
+
+		// found valid header, parse payload
+		tfa303221->message = json_mkobject();
+
+		// add sensor data to json
+		int sensorId = binToDecRev(buffer, 0, 7);
+		int battery = binToDecRev(buffer, 8, 8);
+		int manual = binToDecRev(buffer, 9, 9);
+		int channel = binToDecRev(buffer, 10, 11) + 1;	// bits 0..2, description 1..3
+		int tempRaw = binToDecRev(buffer, 12, 23);
+		float tempCelsius = (tempRaw - 500) * 0.1f;
+		int humidity = binToDecRev(buffer, 24, 31);
+
+// add debug data if enabled
+#ifdef DEBUG
+		json_append_member(tfa303221->message, "rawlen", json_mknumber(tfa303221->rawlen, 0));
+		json_append_member(tfa303221->message, "headerIndex", json_mknumber(payloadStartIndex, 0));
+
+		char *result = malloc(tfa303221->rawlen * 5);
+		char *where = result;
+		for (size_t i = 0; i < tfa303221->rawlen; i++) {
+			size_t printed = sprintf(where, "%d ", tfa303221->raw[i]);
+			where += printed;
+		}
+		json_append_member(tfa303221->message, "raw", json_mkstring(result));
+		free(result);
+
+		char *result1 = malloc(40 * 3);
```

Comment:
You can just reuse `result` and `where` here since `result` has already been freed before.

Reply:
good point, will do